src/Halogen/Canvas/Game.purs

```diff
@@ -0,0 +1,112 @@
+module Halogen.Canvas.Game
+  ( component
+  ) where
+
+import Prelude
+
+import Control.Monad.Reader (runReaderT)
+import Control.Monad.Rec.Class (class MonadRec)
+import Data.Maybe (Maybe(..))
+import Data.Traversable (traverse_)
+import Effect.Aff.Class (class MonadAff)
+import Graphics.Canvas (getContext2D)
+import Graphics.Canvas.Free (CanvasContext, CanvasT, clearRect, getHeight, getWidth, runCanvasT, withContext)
+import Halogen as H
+import Halogen.Canvas.Interact (InputEvent(..), stopInputEventPropagation)
+import Halogen.HTML as HH
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+import Halogen.Subscription as HS
+import Unsafe.Coerce (unsafeCoerce)
+import Web.DOM.Element (DOMRect, getBoundingClientRect)
+import Web.HTML (window)
+import Web.HTML.Window.AnimationFrame (DOMHighResTimestamp, requestAnimationFrame)
+
+type Dimensions =
+  { width :: Int
+  , height :: Int
+  }
+
+
+type Game world m =
+  { dimensions :: Dimensions 
+  , world :: world
+  , draw :: world -> CanvasT m Unit
+  , input :: InputEvent -> DOMRect -> world -> world
+  , animate :: DOMHighResTimestamp -> world -> world
+  }
```

Comment:
Game is a component with a `world` state that has a pure transition function responding to input and a pure transition function responding to animation frames.  The notion is that this is [similar to a Gloss event loop (specifically "play") except the picture is drawn in an imperative style in `CanvasT`](https://hackage.haskell.org/package/gloss-1.13.2.2/docs/Graphics-Gloss.html).